### PR TITLE
feat: add sonic json support

### PR DESCRIPTION
**Pull request by RainshawGao**
Introduce a new drop-in json enhancement: [sonic](https://github.com/bytedance/sonic).

### CODE REVIEW

**Comment by Bisstocuz** [comment]
> ## Requirement
> * Go 1.15/1.16/1.17/1.18
> * Linux/MacOS/Windows
> * Amd64 CPU with AVX instruction set

> Sonic DOSE NOT ensure to support all environments, due to the difficulty of developing high-performance codes. For developers who use sonic to build their applications in different environments (ex: developing on M1 Mac but running on linux server), or those who want to handle JSON strictly consistent with encoding/json, we provide some compatible APIs as sonic.API

It seems not completely drop-in.

**Comment by RainshawGao** [comment]
> > ## Requirement
> > 
> > * Go 1.15/1.16/1.17/1.18
> > * Linux/MacOS/Windows
> > * Amd64 CPU with AVX instruction set
> 
> > Sonic DOSE NOT ensure to support all environments, due to the difficulty of developing high-performance codes. For developers who use sonic to build their applications in different environments (ex: developing on M1 Mac but running on linux server), or those who want to handle JSON strictly consistent with encoding/json, we provide some compatible APIs as sonic.API
> 
> It seems not completely drop-in.

IMO we provide the ability and whether to use should depend on users. we can add more statement in readme.
now if user would like to use sonic, they must have read its docs and know its limitations, this should be supported by our framework

File: internal/json/sonic.go [hunk 1]
```diff
@@ -0,0 +1,20 @@
+//go:build sonic
```
Comment (thinkerou) [inline internal/json/sonic.go:1]: please add copyright, thanks!

**Comment by thinkerou** [comment]
@RainshawGao please update ci, see #3214

File: internal/json/sonic.go (hunk 1 above)
Comment (thinkerou) [reply internal/json/sonic.go:5]: sorry, 2021 -> 2022

**Comment by thinkerou** [comment]
unit test error on sonic tag.

**Comment by Bisstocuz** [comment]
> > 
> > It seems not completely drop-in.
> 
> IMO we provide the ability and whether to use should depend on users. we can add more statement in readme. now if user would like to use sonic, they must have read its docs and know its limitations, this should be supported by our framework

The better solution is for us to identify those incompatible scenarios and then provide compatibility code to ensure the ease of use of the Gin framework. You should add some `build tag` for available hardware architecture.

File: internal/json/sonic.go (hunk 1 above)
Comment (RainshawGao) [reply internal/json/sonic.go]: > sorry, 2021 -> 2022

done

**Comment by RainshawGao** [comment]
> > > It seems not completely drop-in.
> > 
> > 
> > IMO we provide the ability and whether to use should depend on users. we can add more statement in readme. now if user would like to use sonic, they must have read its docs and know its limitations, this should be supported by our framework
> 
> The better solution is for us to identify those incompatible scenarios and then provide compatibility code to ensure the ease of use of the Gin framework. We should add some `build tag` for available hardware architecture.

done, but the fail ci test should be considered more....  https://github.com/golang/go/issues/25956 indicates whether one failed test in ci should be designed or modified still needs discussion.

**Comment by RainshawGao** [comment]
I dig into the fail test cases, all these cases are about trying to use `newdecoder().decode` on an empty string, sonic would not return EOF error while encoding/json does.  Related issues https://github.com/golang/go/issues/25956 

Should this behavior must be the same?

**Comment by RainshawGao** [comment]
@Bisstocuz @thinkerou  sonic has fixed the fail tests, see https://github.com/RainshawGao/gin/pull/1 with action https://github.com/RainshawGao/gin/actions/runs/2607644458 
I have no idea with the error showed in the action summary page as https://github.com/gin-gonic/gin/actions/runs/2565859221 this action also have them.

File: internal/json/json.go [hunk 2]
```diff
@@ -2,8 +2,10 @@
 // Use of this source code is governed by a MIT style
 // license that can be found in the LICENSE file.
 
-//go:build !jsoniter && !go_json
-// +build !jsoniter,!go_json
+//go:build !jsoniter && !go_json && !(sonic && avx && (linux || windows || darwin) && amd64 && (go1.15 || go1.16 || go1.17 || go1.18))
```
Comment (thinkerou) [inline internal/json/json.go:5]: why set go version?

Comment (RainshawGao) [reply internal/json/json.go:5]: @thinkerou as describe by @Bisstocuz, I add the all limits of sonic with build tags, if you think this is no use, I'd like to delete them.

Comment (thinkerou) [reply internal/json/json.go:5]: because `go: [1.15, 1.16, 1.17, 1.18]` has limited, so we should remove it.

File: Makefile [hunk 3]
```diff
@@ -11,7 +11,7 @@ TESTTAGS ?= ""
 test:
 	echo "mode: count" > coverage.out
 	for d in $(TESTFOLDER); do \
-		$(GO) test -tags $(TESTTAGS) -v -covermode=count -coverprofile=profile.out $$d > tmp.out; \
+		$(GO) test $(TESTTAGS) -v -covermode=count -coverprofile=profile.out $$d > tmp.out; \
```
Comment (thinkerou) [inline Makefile:14]: `$(GO) test -tags "$(TESTTAGS)" -v ...`
right?

Comment (RainshawGao) [reply Makefile:14]: oh, I try to fix the error showed in the action summary page, and move the `-tags` string to the action settings, but it seems not work..
```
test-tags: ['', '-tags nomsgpack', '-tags "sonic avx"', '-tags go_json']
```
should this be moved back?

**Review by thinkerou — DISMISSED** [review]

**Comment by thinkerou** [comment]
wait @appleboy approval, thanks!

**Comment by Bisstocuz** [comment]
Good job! But, I think `avx` tag is unneccessary, simply keeping `sonic` is perfect. We only need to ensure developers know about `sonic`'s requirements, and make their Gin service stable if they mistakely use `sonic`.

**Review by thinkerou — APPROVED** [review]

**Review by thinkerou — APPROVED** [review]

File: README.md [hunk 4]
```diff
@@ -204,6 +204,11 @@ go build -tags=jsoniter .
 ```sh
 go build -tags=go_json .
 ```
+[sonic](https://github.com/bytedance/sonic) (you have to ensure that your cpu support avx instruction.)
```
Comment (appleboy) [inline README.md:207]: Add blank line above.

Comment (RainshawGao) [reply README.md:208]: done

**Review by thinkerou — APPROVED** [review]

**Pull request merged**